Add explicit Godunov type solid model

khoderAlshaar opened the pull request:
I have started by adding a new class in **solidModels** named "**solidFoam**," which is consistent with the original solver name in the expleciteSolidDynamics repo.

I have tested the new class with the plotHole tutorial, and it runs with no errors.  

I will be updating the class to match the original solver. 

Besides the solver in the application folder, I need to add the src folder, which contains the boundaryConditions, mathematics,  models, and schemes folders. Please suggest to which places I should add these classes.

khoderAlshaar committed:
adds solidFoam calss for solidModels

philipcardiff commented:
I suggest you add the new src directory as a sub-directory of the solidFoam directory.
Possibly, we could also rename solidFoam to be more descriptive e.g. explicitGodunov..., or something like that.

philipcardiff commented on src/solids4FoamModels/solidModels/solidFoam/solidFoam.H:
```diff
@@ -0,0 +1,144 @@
+    The stress is calculated by the run-time selectable mechanical law.
+
+Author
+    Philip Cardiff, UCD.  All rights reserved.
```
Of course, when you copy the code from solidFoam, make sure to include the original authors.

khoderAlshaar committed:
changes solver name  & adds and compiles sub-directories

khoderAlshaar committed:
Integrate explicitSolidDynamics toolkit  as explicitGodunovCCSolid solidModel

This is a complete integration of the cell centered explicitSolidDynamics toolkit  as a solidModel subclass named explicitGodunovCCSolid.
However, explicitSolidDynamics solves for linear momentum and deformation gradient fields while solids4Foam solves for displacement field.
The next step is to find relations between the fields, and replace the dependencies on the original explicitSolidDynamics libraries with the ones that are available in solids4Foam.

cantilever beam bending test case is added to test the results with solids4Foam with the new solid model.

khoderAlshaar commented on src/solids4FoamModels/solidModels/explicitGodunovCCSolid/createFields.H:
unnecessaryfile. to be deleted.

khoderAlshaar commented on src/solids4FoamModels/solidModels/explicitGodunovCCSolid/readControls.H:
unnecessaryfile. to be deleted.

khoderAlshaar commented:
I have pushed new changes to the repo. I have added the explicitGodunovCC solidModel. The new model solves for linear momentum and deformation gradient. Then, it evaluate the 1st Piola-Kirchhoff stress tensor and total displacement. However, I am facing some difficulties when it comes to solids4Foam fields:

1-  When evaluating the total displacement and stress fields, I take inspiration from explicitUnsLinGeomTotalDisp solidModel. However, the results of the total displacement field are not exactly similar, and the stress is completely wrong. 

2- explicitSolidDynamics solver does not have an option for "empty" type boundary conditions. 

3- The explicitGoduniveCC solid model is slower compared with other models.

4- I am unsure how to choose the damping coefficient value as it affects the convergence speed. I try multiple values with small mesh sizes to find the best one.

philipcardiff commented:
Hi @khoderAlshaar,

What test case are you using to verify the implementation? Knowing the correct answer helps determine whether the implementation is correct. Note that the explicitUnsLinGeomTotalDisp model tends to suffer from stability problems, and I plan to replace it. The linGeomTotalDispSolid is the most tested and is probably better to compare with.

In terms of speed, is it twice as slow (OK) or 10/100/1000 times slower (something is wrong in this case)?

` I am unsure how to choose the damping coefficient value as it affects the convergence speed`: what do you mean by convergence speed? Do you mean convergence to a steady-state?

khoderAlshaar committed:
modify explicitGodunovCCSolid cauch stess update. Add explicitSolidTractionTraction and explicitSolidTractionLinearMomentum BCs. Modify solidModel setTraction.

khoderAlshaar committed:
Merge branch 'development' into new-solid-model

khoderAlshaar committed:
"Modified explicitGodunovCCSolid model: updated fluidSolidInterface, explicitGodunovCCSolid, and mechanics classes; added pesudoTime function; modified strongBCs updated solidModel and solidModelTemplates classes."

khoderAlshaar committed:
Modifies explicitGodunovCCSolid.C and .H files, and changgs gEqns.H and mechanics.C and .H files. works for steady state beamBending.

khoderAlshaar committed:
Adds dual time stepping . works for dynamics problems.

khoderAlshaar committed:
adds condition to work with "empty" BCs. cleans D() calculations and unnecessary calls.

khoderAlshaar committed:
Corrects run in parallel, corrects angular momentum algorithm,  reduces dependcy on old directories, cleans unnecessary code

khoderAlshaar committed:
cleans up fiels from unnecessry lines.

khoderAlshaar committed:
Adds perpendicularFlap and explicitGodunovCCCantilever2d tutorials with explicitGodunoveCC solid model.

khoderAlshaar commented on src/solids4FoamModels/solidModels/solidModel/solidModel.H:
```diff
@@ -316,6 +316,27 @@ protected:
             const bool writeResiduals = true
         );
 
+        //- Check if a vol field  has converged
```
to be cleaned in the next commit

philipcardiff commented:
> Dear @philipcardiff, Adding the explicitGodunoveCC solid model has been completed successfully. I have added two tutorials to run the new solid model: explicitGodunovCCCantilever2d to verify the solid case and perpendicularFlap to verify the FSI case.
> 
> Ref: [preCICE tutorial perpendicular flap](https://precice.org/tutorials-perpendicular-flap.html)
> 
> ![image](https://private-user-images.githubusercontent.com/105800730/389160410-3c24ca80-0143-41d4-8249-0e818f8cd055.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1MzE5NTUsIm5iZiI6MTczMjUzMTY1NSwicGF0aCI6Ii8xMDU4MDA3MzAvMzg5MTYwNDEwLTNjMjRjYTgwLTAxNDMtNDFkNC04MjQ5LTBlODE4ZjhjZDA1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyNVQxMDQ3MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNTZlYTc3ZjkzMDc3OGZmMTRiYmNkMTE1MTdkNTM5M2IxODNiNzc4ODExNjdlMmY0YTRlNDk1Y2ZiOTQzYmNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ZFb58XxU_NdMOSNUtZLAB30-dk6xWyEi1ySZpuUg6l8) ![image](https://private-user-images.githubusercontent.com/105800730/389057447-dc71aee1-234b-48dc-9eef-15e64525d52f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1MzE5NTUsIm5iZiI6MTczMjUzMTY1NSwicGF0aCI6Ii8xMDU4MDA3MzAvMzg5MDU3NDQ3LWRjNzFhZWUxLTIzNGItNDhkYy05ZWVmLTE1ZTY0NTI1ZDUyZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyNVQxMDQ3MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZWYyZTFkMTg0Y2NhMzMwZjUxMTcyODYyODMwMTFmNTg5YjE0M2JkMmI2NjQyZDYyOGNhZmUxZTU2MTIwNjY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.tWj-0i5FxZExcKlYN2weyYcVcNNoBNW3OkBG4yw7e3I)

Thanks, @khoderAlshaar, this looks impressive. Apologies for my delayed commenting on this. I am in the last week of the teaching term and hope to have more time in the coming weeks. I will then review the code, make suggestions (if needed), and merge it.

What is your planned next step with this?

philipcardiff commented:
FYI, see the auto-checks above too.

khoderAlshaar commented:
> > Dear @philipcardiff, Adding the explicitGodunoveCC solid model has been completed successfully. I have added two tutorials to run the new solid model: explicitGodunovCCCantilever2d to verify the solid case and perpendicularFlap to verify the FSI case.
> > Ref: [preCICE tutorial perpendicular flap](https://precice.org/tutorials-perpendicular-flap.html)
> > ![image](https://private-user-images.githubusercontent.com/105800730/389160410-3c24ca80-0143-41d4-8249-0e818f8cd055.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1MzE5NTUsIm5iZiI6MTczMjUzMTY1NSwicGF0aCI6Ii8xMDU4MDA3MzAvMzg5MTYwNDEwLTNjMjRjYTgwLTAxNDMtNDFkNC04MjQ5LTBlODE4ZjhjZDA1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyNVQxMDQ3MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNTZlYTc3ZjkzMDc3OGZmMTRiYmNkMTE1MTdkNTM5M2IxODNiNzc4ODExNjdlMmY0YTRlNDk1Y2ZiOTQzYmNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ZFb58XxU_NdMOSNUtZLAB30-dk6xWyEi1ySZpuUg6l8) ![image](https://private-user-images.githubusercontent.com/105800730/389057447-dc71aee1-234b-48dc-9eef-15e64525d52f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1MzE5NTUsIm5iZiI6MTczMjUzMTY1NSwicGF0aCI6Ii8xMDU4MDA3MzAvMzg5MDU3NDQ3LWRjNzFhZWUxLTIzNGItNDhkYy05ZWVmLTE1ZTY0NTI1ZDUyZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyNVQxMDQ3MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZWYyZTFkMTg0Y2NhMzMwZjUxMTcyODYyODMwMTFmNTg5YjE0M2JkMmI2NjQyZDYyOGNhZmUxZTU2MTIwNjY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.tWj-0i5FxZExcKlYN2weyYcVcNNoBNW3OkBG4yw7e3I)
> 
> Thanks, @khoderAlshaar, this looks impressive. Apologies for my delayed commenting on this. I am in the last week of the teaching term and hope to have more time in the coming weeks. I will then review the code, make suggestions (if needed), and merge it.
> 
> What is your planned next step with this?

Thank you for taking the time to review the code @philipcardiff.  I am working on this as a final project for the OSCFD course at Chalmers University. 

About my next step, I have multiple ideas in mind. The main goal is to test the explicit model against  FSI problems with compressible flow. I am also considering adding a compressible fluid model from the dbns family. 

However, I am still testing the performance of the new explicitGodonov model. Accuracy looks acceptable for linear problems, but the CPU time is large in certain cases. So, I would like to address this issue first and see what parts need improvements.

khoderAlshaar committed:
test commit

khoderAlshaar committed:
Modifies license for all files

khoderAlshaar committed:
Corrects  license for all files

khoderAlshaar committed:
Uses solids4FoamChangeCopyright.sh  to correct License

philipcardiff committed:
Merge branch 'development' into new-solid-model

philipcardiff commented:
Hi @khoderAlshaar, before merging, we will need to address the build failures above, e.g. for OF-9 and/or FE41. If we can't support those versions, we can always disable the functionality in the code with compiler directives, e.g. `#ifdef OPENFOAM_COM`

khoderAlshaar commented:
Hi @philipcardiff, I want to update the code so it works with other OF versions. However, due to time restrictions, I would like to keep the current version to run with OFv2012. 
At the moment, I want to extend the explicit Godunov solid model by adding a runTimeSelction mechanism to choose from a set of available Riemann solvers that I am currently developing.

khoderAlshaar committed:
Redefines setTraction function in explicitGodunovCCSolid Class

khoderAlshaar committed:
Merge branch 'new-solid-model' of https://github.com/khoderAlshaar/solids4foam into new-solid-model
